File: day1.py
#day1.py 2021
import unittest

def part1(readings):
    last = None
    inc = 0
    for reading in [int(r) for r in readings]:
        if last is not None and reading > last:
            inc += 1
        last = reading
    return inc
        

def part2(readings, window=3):
    last = None
    inc = 0
    readings = [int(r) for r in readings]
    for i in range(len(readings)-window+1):
        reading = sum(readings[i:i+window])
        if last is not None and reading > last:
            inc += 1
        last = reading
    return inc

class TestDay1(unittest.TestCase):
    def test_1a(self):
        self.assertEqual(part1([199,200,208,210,200,207,240,269,260,263]), 7)

    def test_1(self):
        with open('./input1.txt', 'r') as f:
            self.assertEqual(part1(f), 1713)

    def test_2a(self):
        self.assertEqual(part2([199,200,208,210,200,207,240,269,260,263]), 5)

    def test_2b(self):
        self.assertEqual(part2([199,200,208,210,200,207,240,269,260,263], 1), 7)
    
    def test_1(self):
        with open('./input1.txt', 'r') as f:
            self.assertEqual(part2(f), 1734)

if __name__ == '__main__':
    unittest.main()

File: cpp/day1.cpp
// day1.cpp
#include <gtest/gtest.h>
#include <string>
#include <fstream>
#include <numeric>
using namespace std;

int part1(ifstream &f)
{
    string line;
    int last = INT_MAX;
    int increases = 0;
    while (getline (f,line))
    {
        int x = stoi(line);
        if (x > last) {
            increases++;
        }
        last = x;
    }
    return increases;
}

int part2(ifstream &f, int window=3)
{
    vector<int> v;
    string line;
    while (getline (f,line))
    {
        v.push_back(stoi(line));
    }

    int last = INT_MAX;
    int inc = 0;
    for (int i=0; i<v.size()-window+1; i++) {
        int reading = accumulate(v.begin() + i, v.begin() + i + window, 0);
        if (reading > last) {
            inc++;
        }
        last = reading;
    }
    
    return inc;
}

TEST(Day1, Part1_test) {
    ifstream infile("../test1.txt");
    ASSERT_EQ(part1(infile), 7);
    infile.close();
}

TEST(Day1, Part1) {
    ifstream infile("../input1.txt");
    ASSERT_EQ(part1(infile), 1713);
    infile.close();
}

TEST(Day1, Part2_test) {
    ifstream infile("../test1.txt");
    ASSERT_EQ(part2(infile), 5);
    infile.close();
}

TEST(Day1, Part2) {
    ifstream infile("../input1.txt");
    ASSERT_EQ(part2(infile), 1734);
    infile.close();
}
